10up/classifai
# Add support for role/user based access

### Description of the Change
This PR adds support for role- and user-based access control to the ClassifiAI feature. It also includes functionality for user-based opt-out, allowing users to choose not to use the specific feature if they prefer not to or if they dislike it.

**Feature settings:**
<img width="877" alt="image" src="https://github.com/10up/classifai/assets/10613171/de6b810b-b070-4751-86e6-41738cf9022a">

**User Profile:**
<img width="1020" alt="image" src="https://github.com/10up/classifai/assets/10613171/ba00f8fb-0c34-4d87-8794-8e724a14a44e">

**Edit User:**
<img width="1020" alt="image" src="https://github.com/10up/classifai/assets/10613171/fce99eff-5164-4aab-9bb3-0e3344dd4c74">

**In-Context disable feature link to user profile**
|Title generation|Resize content|
|---|---|
|![Screenshot 2023-10-30 at 5 17 15 PM](https://github.com/10up/classifai/assets/10613171/31daa294-61ac-45e4-b635-d579acbfaf60)|![Screenshot 2023-10-30 at 5 18 06 PM](https://github.com/10up/classifai/assets/10613171/aa9b63b7-c4a2-4419-98df-0ac9ab04dad3)|

_**Note:** E2E tests are not added for this new feature._


Closes #588

### How to test the Change
1. Enable role-based access to the feature.
1. Select roles in the allowed roles to provide access to that feature.
1. Verify that only the selected roles have access to that feature.
1. Enable user-based access to the feature.
1. Select the user in the allowed users to provide access to that feature.
1. Verify that the selected user has access to that feature.
1. Enable user-based opt-out for the feature.
1. Go to the user profile and verify that you see the checkbox for that feature to opt out of using it.
1. Check the opt-out checkbox and save.
1. Verify that access to that feature has been disabled.
1. Uncheck the opt-out checkbox to opt back in to the feature.
1. Verify that the user now has access to that feature.
1. Follow these steps for each ClassifAI feature.


### Changelog Entry
> Added - Role and user-based access control for features


### Credits
Props @jeffpaul @dkotter @iamdharmesh @Sidsector9 


### Checklist:
